### PR TITLE
What is a particle?

### PR DESCRIPTION
**Pull request by nahueespinosa**
### Proposed changes

Meta-programming magic to better define the particle concept and constraints (with tests!).

Related to #279.

#### Type of change

- [ ] 🐛 Bugfix (change which fixes an issue)
- [x] 🚀 Feature (change which adds functionality)
- [ ] 📚 Documentation (change which fixes or extends documentation)

### Checklist

- [x] Lint and unit tests (if any) pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] All commits have been signed for [DCO](https://developercertificate.org/)

### CODE REVIEW

File: beluga/include/beluga/primitives.hpp [hunk 1]
```diff
@@ -67,27 +67,27 @@ struct has_member_state : std::false_type {};
...
+/// \endcond
 
 /// Customization point object type for accessing the `state` of a particle.
 /**
  * See https://en.cppreference.com/w/cpp/ranges/cpo.
  */
 struct state_fn {
   /// Overload for when the particle type defines a member method.
-  template <typename T, std::enable_if_t<has_member_state_v<T>, int> = 0>
+  template <class T, std::enable_if_t<has_member_state_v<T>, int> = 0>
```
Comment (hidmic) [inline beluga/include/beluga/primitives.hpp:90]: @nahueespinosa meta: why support getter methods but not plain members? A `struct` with `state` and `weight` members is the simplest particle definition I can think of, even if it works against performance.

File: beluga/include/beluga/primitives.hpp [hunk 2]
```diff
@@ -67,27 +67,27 @@ struct has_member_state : std::false_type {};
...
   constexpr decltype(auto) operator()(T&& t) const noexcept(noexcept(std::forward<T>(t).state())) {
```
Comment (hidmic) [inline beluga/include/beluga/primitives.hpp:91]: @nahueespinosa TIL building `noexcept` specifiers using `noexcept` operators is a thing.

**Review by hidmic — COMMENTED** [review]
This is amazing. Only one comment / wish.

File: beluga/include/beluga/primitives.hpp (hunk 1 above)
Comment (nahueespinosa) [reply beluga/include/beluga/primitives.hpp:90]: You got it! I'll update the PR soon.

Comment (nahueespinosa) [reply beluga/include/beluga/primitives.hpp]: Done in https://github.com/Ekumen-OS/beluga/pull/280/commits/5db0b4084d498269ed7d9d967212e7c2e6f88426!

**Review by hidmic — APPROVED** [review]
Looks awesome to me.

File: beluga/include/beluga/primitives.hpp [hunk 3]
```diff
@@ -61,33 +62,54 @@ namespace state_detail {
...
+
+/// \endcond
 
 /// Customization point object type for accessing the `state` of a particle.
 /**
  * See https://en.cppreference.com/w/cpp/ranges/cpo.
  */
 struct state_fn {
+  /// Overload for when the particle type defines a member variable.
+  template <class T, std::enable_if_t<has_member_variable_state_v<T>, int> = 0>
```
Comment (hidmic) [inline beluga/include/beluga/primitives.hpp:100]: @nahueespinosa meta: it occurs to me that, perhaps, making overloads mutually exclusive may spare someone some pain (_why is ADL resolution not working? oooh, I have an `state` member... -.-_)

Comment (nahueespinosa) [reply beluga/include/beluga/primitives.hpp]: Makes sense to me, done in https://github.com/Ekumen-OS/beluga/pull/280/commits/2048b4e91f30abe3b8a987735e4f741671cef254.

**Pull request merged**